Hardcode debian sendmail path
--- a/configure.ac
+++ b/configure.ac
@@ -222,7 +222,7 @@
   --with-sendmail=PATH    Specify path to sendmail,
             [dnl
 case "$with_sendmail" in
-yes|''|no) AC_MSG_ERROR([--with-sendmail   requires an argument; use --with-sendmail=PATH]) ;;
+yes|no) AC_MSG_ERROR([--with-sendmail   requires an argument; use --with-sendmail=PATH]) ;;
 *) SENDMAIL="$withval" ;;
 esac
 ])
@@ -1114,15 +1114,7 @@
 
 # Checking sendmail
 AC_MSG_CHECKING(for sendmail)
-if test "${with_sendmail+set}" = set ;then
-	SENDMAIL="$with_sendmail"
-else
-    XPATH="/usr/sbin:/usr/lib:/usr/ucblib:/usr/share"
-    AC_PATH_PROG(SENDMAIL, sendmail,, $XPATH)
-    if test "x$SENDMAIL" = "x"; then
-	AC_MSG_ERROR([sendmail not found, re-run with valid --with-sendmail=DIR])
-    fi
-fi
+SENDMAIL = "/usr/sbin/sendmail"
 AC_SUBST(SENDMAIL)
 
 AC_MSG_CHECKING(for lock directory)
