Fix debian news pathes
--- a/configure.ac
+++ b/configure.ac
@@ -1083,16 +1083,16 @@
     fi
 }
 else
-    BATCHER='time $RFC2FTN -f $BATCHFILE -m 500'
-    RNEWS='/none/rnews'
-    INEWS='/none/inews'
-    NEWSLIBDIR='/none/news-libdir'
-    PATHOUTGOING='/none/news-pathoutgoing'
-    PATHARTICLES='/none/news-patharticles'
-    PATHNEWS='/none/news-pathnews'
-    PATHBIN='/none/news-pathbin'
-    PATHETC='/none/news-pathetc'
-    PATHDB='/none/news-pathdb'
+    BATCHER='time $PATHBIN/batcher -N $QUEUEJOBS -b500000 -p"$RFC2FTN -b -n" $SITE $BATCHFILE'
+    RNEWS='/usr/lib/news/bin/rnews'
+    INEWS='/usr/lib/news/bin/inews'
+    NEWSLIBDIR='/usr/lib/news'
+    PATHOUTGOING='/var/spool/news/outgoing'
+    PATHARTICLES='/var/spool/news/articles'
+    PATHNEWS='/usr/lib/news'
+    PATHBIN='/usr/lib/news/bin'
+    PATHETC='/etc/news'
+    PATHDB='/var/lib/news'
 fi
 
 AC_SUBST(BATCHER)
